pkg/osquery/tables/xfconf/xfconf.go

```diff
@@ -0,0 +1,164 @@
+//go:build !windows
+// +build !windows
+
+package xfconf
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"os"
+	"os/user"
+	"path/filepath"
+	"strings"
+
+	"github.com/go-kit/kit/log"
+	"github.com/kolide/launcher/pkg/dataflatten"
+	"github.com/kolide/launcher/pkg/osquery/tables/dataflattentable"
+	"github.com/kolide/launcher/pkg/osquery/tables/tablehelpers"
+	"github.com/osquery/osquery-go/plugin/table"
+)
+
+// Provides configuration settings for devices using XFCE desktop environment.
+// See https://docs.xfce.org/xfce/xfce4-session/advanced#files_and_environment_variables
+// for a list of the files and environment variables we check for configuration.
+
+var xfconfChannelXmlPath string = filepath.Join("xfce4", "xfconf", "xfce-perchannel-xml")
+
+type XfconfQuerier struct {
+	logger log.Logger
+}
+
+func TablePlugin(logger log.Logger) *table.Plugin {
+	t := &XfconfQuerier{
+		logger: logger,
+	}
+
+	return table.NewPlugin("kolide_xfconf", dataflattentable.Columns(table.TextColumn("username"), table.TextColumn("channel")), t.generate)
+}
+
+func (t *XfconfQuerier) generate(ctx context.Context, queryContext table.QueryContext) ([]map[string]string, error) {
+	var results []map[string]string
+
+	users := tablehelpers.GetConstraints(queryContext, "username")
+	if len(users) < 1 {
+		return results, errors.New("kolide_xfconf requires at least one username to be specified")
+	}
+
+	for _, username := range users {
+		u, err := user.Lookup(username)
+		if err != nil {
+			return nil, fmt.Errorf("finding user by username '%s': %w", username, err)
+		}
+
+		rowData := map[string]string{"username": username}
+
+		for _, dataQuery := range tablehelpers.GetConstraints(queryContext, "query", tablehelpers.WithDefaults("*")) {
+			userConfig, err := t.getUserConfig(u, dataQuery, rowData)
+			if err != nil {
+				return nil, fmt.Errorf("could not get xfconf settings for user %s and query %s: %w", username, dataQuery, err)
+			}
+
+			results = append(results, userConfig...)
+		}
+	}
+
+	return results, nil
+}
+
+func (t *XfconfQuerier) getUserConfig(u *user.User, dataQuery string, rowData map[string]string) ([]map[string]string, error) {
+	var results []map[string]string
+
+	// First, get user-specific settings
+	userConfigDir := getUserXfconfDir(u)
+	userConfigRows, err := t.getConfigFromDirectory(userConfigDir, dataQuery, rowData)
+	if err != nil {
+		return nil, fmt.Errorf("error getting config for user %s from directory %s: %w", u.Name, userConfigDir, err)
+	}
+	results = append(results, userConfigRows...)
+
+	// Then, get defaults
```

Comment:
Defaults won't change during the course of a `generate` call, but we will likely invoke for multiple users. This should, ideally, move out of the `getUserConfig` so we don't regen each time.

Reply:
Oh yup, good suggestion, I will move this and https://github.com/kolide/launcher/pull/948/#discussion_r1036632898 up so we don't call it for every user